csrc/device_lower/pass/index.cpp

```diff
@@ -1451,17 +1451,18 @@ static Val* constructMatrixDescriptor(
 }
 
 static MmaInputSmemSwizzle getSwizzleMode(TensorView* tv) {
-  auto n_alloc = tv->getAllocationDomain().size();
-  auto num_core_matrices = tv->getAllocationDomain()
-                               .at(n_alloc - 2)
-                               ->extent()
-                               ->evaluate()
-                               .as<int64_t>() /
-      8;
-  if (num_core_matrices == 1) {
-    return MmaInputSmemSwizzle::None;
+  const auto& alloc_domain = tv->getRootDomain();
+  const auto& leaf_domain = tv->getLeafDomain();
+  auto exprs = StmtSort::getExprsBetween(
+      {alloc_domain.begin(), alloc_domain.end()},
+      {leaf_domain.begin(), leaf_domain.end()});
+  for (auto expr : exprs) {
+    if (auto swizzle = dynamic_cast<const Swizzle*>(expr)) {
```

Comment:
Do we plan to handle anyhow situations where by mistake there are two swizzle ops applied on one of mma inputs?

Currently we search for first instance, and return its swizzle, while we could put a strong requirement to have a single instance in the expr set. Something like this:

```
std::vector<const Swizzle*> swizzle_exprs = [](){
  std::vector<const Swizzle*> results;
  for (auto expr: exprs) {
    if (auto swizzle = dynamic_cast<const Swizzle*>(expr)) {
      results.push_back(swizzle);
    }
  }
  return results;
}();
NVF_ERROR(swizzle_exprs.size() < 2, "expected 2 or less swizzle expressions in mma input, got ", swizzle_exprs.size());

if (swizzle_exprs.empty()) {
  swizzle_exprs.front()->inX()->extent()->evaluate().as<int64_t>() * 16);
}
else {
  return MmaInputSmemSwizzle::None;
}
```

The next step could be checking if there are more than a single swizzle expr but all of them have the same type of swizzle (effectively may mean that second and following exprs could be no-ops)

Reply:
Added

Reply:
Your implementation looks better than the suggestion - I learnt something new. Thanks!